server/src/main/java/org/elasticsearch/search/query/TopDocsCollectorContext.java

```diff
@@ -357,6 +364,20 @@ static int shortcutTotalHitCount(IndexReader reader, Query query) throws IOExcep
                 count += context.reader().docFreq(term);
             }
             return count;
+        } else if (query.getClass() == DocValuesFieldExistsQuery.class && reader.hasDeletions() == false) {
```

Comment:
Would it make sense to add a check for `NormsFieldExistsQuery`, so that the optimization can be applied to `text` fields as well?

Reply:
Would the same pre-requisites apply (no deletion, indexed) and would the count also be derived from terms.getDocCount() in that case?

Reply:
Unfortunately it doesn't work because `Terms#getDocCount` only counts documents that have at least one token while norms record a value of 0 for documents that have a field value without terms. So if you were to index text fields that don't have any tokens, such as empty strings or strings that are only made of stopwords, you would get wrong counts.

Reply:
Thanks for the explanation!